src/main/java/me/jellysquid/mods/sodium/client/events/RenderBlockCallback.java

```diff
@@ -0,0 +1,34 @@
+package me.jellysquid.mods.sodium.client.events;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.minecraft.block.BlockState;
+import net.minecraft.util.math.Direction;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Allows taking control of culling/occlusion yourself. This should only be done where other methods are not
+ * possible, an example of this is forge/porting lib's obj loader that doesn't support configuring culling
+ * manually.
+ *
+ * <p>
+ * Example: <pre>
+ * {@code
+ * RenderBlockCallback.EVENT.register((state, adjacentBlockState, direction) -> {
+ *  return (adjacentBlockState.is(MyModsBlocks.CAKE_BLOCK.get()) && adjacentBlockState.getValue(HORIZONTAL_AXIS) == state.getValue(HORIZONTAL_AXIS))
+ * });
+ * }
+ * </pre>
+ */
+@FunctionalInterface
+public interface RenderBlockCallback {
+    Event<RenderBlockCallback> EVENT = EventFactory.createArrayBacked(RenderBlockCallback.class,
+            (listeners) -> (state, adjacentBlockState, direction) -> {
+                for (RenderBlockCallback listener : listeners) {
+                    return listener.selfManageOcclusion(state, adjacentBlockState, direction);
```

Comment:
Does it make sense to return directly in this loop if there are multiple listeners. Also, is it performant to run this double lambda and an iterator loop on every block? (it is cached afaik, but still)

Reply:
1. Pretty sure it'll have to be inside the loop since this is whats actually calling the method on the like callback letting x mod do whatever checks/culling they want
2. havent tested but ill try it out right now

Reply:
I meant that only the first listener is called if you return in the loop like that, which I assume isn't intentional.

Reply:
RenderBlockCallback takes up 255ms on my machine which is 0.14% of all

running `RenderBlockCallback.EVENT.register((state, adjacentBlockState, direction) -> (adjacentBlockState.isOf(Blocks.DIAMOND_BLOCK)));` in sodiums oninit takes up 138ms

for the invoker in shouldDrawSide it takes up 586ms which is a little higher then `BlockState adjState = view.getBlockState(adjPos)` line, in total its 24% of shouldDrawSide so it should be fine

(tested with intellij's profiler on linux with nvidia 3060 & i9-9900k)

Reply:
> I meant that only the first listener is called if you return in the loop like that, which I assume isn't intentional.

unless im missing something it should be calling all the listeners as seen in https://fabricmc.net/wiki/tutorial:events but i may be understanding this wrongly

Reply:
after testing yeah your correct only the first listener is called, lemme go fix that